cni/pkg/nftables/host_sets.go

```diff
@@ -0,0 +1,385 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nftables
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+
+	"sigs.k8s.io/knftables"
+
+	"istio.io/istio/tools/istio-nftables/pkg/builder"
+)
+
+// HostNftSetManager implements AddressSetManager using nftables sets
+type HostNftSetManager struct {
+	v4SetName   string
+	v6SetName   string
+	prefix      string
+	enableIPv6  bool
+	nftProvider NftProviderFunc
+}
+
+// NewHostSetManager creates a new nftables-based host set manager
+func NewHostSetManager(setPrefix string, enableIPv6 bool) (*HostNftSetManager, error) {
+	// Lets use the same naming convention as ipset
+	v4SetName := fmt.Sprintf("%s-v4", setPrefix)
+	v6SetName := fmt.Sprintf("%s-v6", setPrefix)
+
+	// Use the real implementation when the package level variable is not set
+	nftProvider := nftProviderVar
+	if nftProvider == nil {
+		nftProvider = func(family knftables.Family, table string) (builder.NftablesAPI, error) {
+			return builder.NewNftImpl(family, table)
+		}
+	}
+
+	manager := &HostNftSetManager{
+		v4SetName:   v4SetName,
+		v6SetName:   v6SetName,
+		prefix:      setPrefix,
+		enableIPv6:  enableIPv6,
+		nftProvider: nftProvider,
+	}
+
+	// Initialize the sets
+	if err := manager.initializeSets(); err != nil {
+		return nil, err
+	}
+
+	return manager, nil
+}
+
+// initializeSets creates the nftables table and sets
+func (h *HostNftSetManager) initializeSets() error {
+	nft, err := h.nftProvider(knftables.InetFamily, AmbientNatTable)
+	if err != nil {
+		return err
+	}
+
+	tx := nft.NewTransaction()
+
+	// Create table
+	tx.Add(&knftables.Table{})
+
+	// Create IPv4 set
+	tx.Add(&knftables.Set{
+		Name:    h.v4SetName,
+		Type:    "ipv4_addr",
+		Comment: knftables.PtrTo("UUID of the pods that are part of ambient mesh"),
+	})
+
+	// Create IPv6 set if enabled
+	if h.enableIPv6 {
+		tx.Add(&knftables.Set{
+			Name:    h.v6SetName,
+			Type:    "ipv6_addr",
+			Comment: knftables.PtrTo("UUID of the pods that are part of ambient mesh"),
+		})
+	}
+
+	if err := nft.Run(context.TODO(), tx); err != nil {
+		return fmt.Errorf("failed to initialize nftables sets: %w", err)
+	}
+
+	builder.LogNftRules(tx)
+	return nil
+}
+
+// AddIP adds an IP address to the appropriate set (v4 or v6)
+func (h *HostNftSetManager) AddIP(ip netip.Addr, ipProto uint8, comment string, replace bool) error {
+	ipToInsert := ip.Unmap()
```

Comment:
So we can never add an ipv6 mapped v4 address to the v6 set (since we unmap and then check the family)? This may be correct, but just want to verify